crates/noirc_evaluator/src/ssa/context.rs

```diff
@@ -1020,16 +1071,122 @@ impl SsaContext {
             let operation = Operation::Phi { root: v, block_args: vec![(a, block1), (b, block2)] };
-            let new_phi = node::Instruction::new(operation, v_type, Some(exit_block));
+            let new_phi = node::Instruction::new(operation, a_type, Some(exit_block));
             let phi_id = self.add_instruction(new_phi);
             self[exit_block].instructions.insert(1, phi_id);
             phi_id
         }
     }
 
+    pub fn push_function_id(&mut self, func_id: FuncId, typ: ObjectType) {
+        let index = self.nodes.insert_with(|index| {
+            let node_id = NodeId(index);
+            NodeObj::Function(FunctionKind::Normal(func_id), node_id, typ)
+        });
+
+        self.function_ids.insert(func_id, NodeId(index));
+    }
+
+    /// Return the standard NodeId for this FuncId.
+    /// The 'standard' NodeId is just the NodeId assigned to the function when it
+    /// is first compiled so that repeated NodeObjs are not made for the same function.
+    /// If this function returns None, it means the given FuncId has yet to be compiled.
+    pub fn get_function_node_id(&self, func_id: FuncId) -> Option<NodeId> {
+        self.function_ids.get(&func_id).copied()
+    }
+
+    pub fn function_already_compiled(&self, func_id: FuncId) -> bool {
+        self.get_ssafunc(func_id).is_some()
+    }
+
+    pub fn get_or_create_opcode_node_id(&mut self, opcode: acvm::acir::OPCODE) -> NodeId {
+        if let Some(id) = self.opcode_ids.get(&opcode) {
+            return *id;
+        }
+
+        let (len, elem_type) = function::get_result_type(opcode);
+        let array_set = if len > 1 {
+            let array = self.new_array(&format!("{}_result", opcode), elem_type, len, None).1;
+            self.push_array_set(vec![(array, 0)])
+        } else {
+            self.push_array_set(vec![])
+        };
+
+        let index = self.nodes.insert_with(|index| {
+            let node_id = NodeId(index);
+            let typ = ObjectType::Function(array_set);
+            NodeObj::Function(FunctionKind::Builtin(opcode), node_id, typ)
+        });
+        self.opcode_ids.insert(opcode, NodeId(index));
+        NodeId(index)
+    }
+
+    pub fn get_builtin_opcode(&self, node_id: NodeId) -> Option<OPCODE> {
+        match &self[node_id] {
+            NodeObj::Function(FunctionKind::Builtin(opcode), ..) => Some(*opcode),
+            _ => None,
+        }
+    }
+
+    pub fn push_array_set(&mut self, arrays: Vec<(ArrayId, u32)>) -> ArrayIdSet {
+        let next_id = self.function_returned_arrays.len();
+        self.function_returned_arrays.push(arrays);
+        ArrayIdSet(next_id as u32)
+    }
+
+    pub fn union_array_sets(&mut self, first: ArrayIdSet, second: ArrayIdSet) -> ArrayIdSet {
+        let mut union_set = self.function_returned_arrays[first.0 as usize].clone();
+        union_set.extend_from_slice(&self.function_returned_arrays[second.0 as usize]);
+        union_set.sort();
+        union_set.dedup();
```

Comment:
I mostly kept it this way because `get_returned_arrays` returns a `Vec<(ArrayId, u32)>` to match what is expected for call instructions. So storing that Vec directly means we wouldn't have to convert into a Vec there, but need to sort and dedup when taking a `Phi` value of two functions which is usually rare.

Reply:
We don't really have to convert the Vec in `get_returned_arrays` with a map though no? We can just have something like: 
```
pub struct SsaContext {
...
    pub function_returned_arrays_map: HashMap<ArrayIdSet, Vec<(ArrayId, /*result_index:*/ u32)>>,
...
}
```
```
    pub fn get_returned_arrays(&self, func: NodeId) -> Vec<(ArrayId, u32)> {
        match self[func].get_type() {
            ObjectType::Function(arrays) => {
                self.function_returned_arrays_map[&arrays].clone()
            }
            other => unreachable!("get_returned_arrays: Expected function type, found {:?}", other),
        }
    }
```
And then we can get rid of the need to sort and dedup at all even if taking a `phi` value of two functions is rare. Lmk if I am missing something here or if this change would break a lot of other things.

If we must keep the sort and dedup can we leave a comment mentioning that because `union_array_sets` is used in `new_phi` it will be rarely called?

Reply:
Ok perfect, that is all clear now, thanks for elaborating.